src/operator.c

```diff
@@ -11,415 +12,1435 @@
+  }
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+// performs all work necessary to make the operator ready for use
+static PetscErrorCode AddSuboperators(Operator *op) {
+  PetscFunctionBegin;
+
+  if (CeedEnabled()) {
+    PetscCall(AddCeedSuboperators(op));
+  } else {
+    PetscCall(AddPetscSuboperators(op));
+  }
+
+  // set up our flux divergence vector(s)
+  PetscCall(AddOperatorFluxDivergence(op));
+
+  // initialize diagnostics
+  PetscCall(ResetOperatorDiagnostics(op));
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/// Creates an operator representing the system of equations described in the
+/// given configuration.
+/// @param [in]  config              the configuration defining the physics and numerics for the new operator
+/// @param [in]  domain_dm           a DM representing the computational domain on which the operator is defined
+/// @param [in]  domain_mesh         a mesh containing geometric and topological information for the domain
+/// @param [in]  num_regions         the number of disjoint regions partitioning the computational domain
+/// @param [in]  regions             an array of disjoint regions paratitioning the computational domain
+/// @param [in]  num_boundaries      the number of distinct boundaries bounding the computational domain
+/// @param [in]  boundaries          an array of distinct boundaries bounding the computational domain
+/// @param [in]  boundary_conditions an array of boundary conditions corresponding to the domain boundaries
+/// @param [out] op                  the newly created operator
+PetscErrorCode CreateOperator(RDyConfig *config, DM domain_dm, RDyMesh *domain_mesh, PetscInt num_regions, RDyRegion *regions,
+                              PetscInt num_boundaries, RDyBoundary *boundaries, RDyCondition *boundary_conditions, Operator **operator) {
+  PetscFunctionBegin;
+
+  MPI_Comm comm;
+  PetscCall(PetscObjectGetComm((PetscObject)domain_dm, &comm));
+
+  // check our arguments
+  PetscCheck(domain_mesh, comm, PETSC_ERR_USER, "Cannot create an operator with no mesh");
+  PetscCheck(num_regions > 0, comm, PETSC_ERR_USER, "Cannot create an operator with no regions");
+  PetscCheck(num_boundaries > 0, comm, PETSC_ERR_USER, "Cannot create an operator with no boundaries");
+
+  static PetscBool first_time = PETSC_TRUE;
+  if (first_time) {
+    // register a logging event for applying our CEED operator
+    PetscCall(PetscLogEventRegister("CeedOperatorApp", RDY_CLASSID, &RDY_CeedOperatorApply_));
+    first_time = PETSC_FALSE;
+  }
+
+  PetscCall(PetscCalloc1(1, operator));
+  (*operator)->config         = config;
+  (*operator)->num_components = 3;
+
+  PetscCall(SetOperatorDomain(*operator, domain_dm, domain_mesh));
+  PetscCall(SetOperatorBoundaries(*operator, num_boundaries, boundaries, boundary_conditions));
+  PetscCall(SetOperatorRegions(*operator, num_regions, regions));
+  PetscCall(AddSuboperators(*operator));
```

Comment:
Should the code here be put within a `switch` the following? 

```c
switch (config->physics.flow) {
  case FLOW_SWE:
   ...

}
```
Or will we do that refactor as a second PR?